### PR TITLE
feature: PortfolioEnv 완전 구현+보상함수 3종 구현+PPO 50k 스텝 테스트 학습

### LINKED ISSUE #12: [FEAT] PortfolioEnv 거래 환경 완전 구현
## 관련 스프린트
- [ ] Sprint 1 (3/31~4/13)
- [x] Sprint 2 (4/14~4/27)
- [ ] Sprint 3 (4/28~5/11)
- [ ] Sprint 4 (5/12~5/25)
- [ ] Sprint 5 (5/26~6/15)

## 기능 설명
Gymnasium 기반 PortfolioEnv의 핵심 거래 로직을 완성한다.  
거래비용, 슬리피지, 포트폴리오 비중 정규화, MDD 기반 Safe-Guard를 포함한다.

## 필요한 이유
PPO 학습이 실제 포트폴리오 환경에서 동작하려면 step/reset 로직과 리스크 제어가 먼저 완성되어야 한다.

## 구현 방향 (선택)
- 행동값을 softmax로 정규화하여 비중 합 1 보장
- 리밸런싱 시 비중 변화량에 비례해 거래비용 0.015% + 슬리피지 0.05% 차감
- 현재 MDD가 15% 초과하면 done=True 처리
- 관측값: 과거 수익률 + 현재 비중 + RSI + MACD

## 관련 파일/모듈 (선택)
수정 또는 추가가 필요한 파일을 적어주세요.
- `src/rl/env.py`
- `tests/test_env.py`

## 담당자
- [ ] 강유영 (RAG/프론트)
- [ ] 박지민 (데이터/백엔드)
- [x] 이문정 (RL/인프라)

### LINKED ISSUE #13: [FEAT] 보상함수 3종 구현
## 관련 스프린트
- [ ] Sprint 1 (3/31~4/13)
- [x] Sprint 2 (4/14~4/27)
- [ ] Sprint 3 (4/28~5/11)
- [ ] Sprint 4 (5/12~5/25)
- [ ] Sprint 5 (5/26~6/15)

## 기능 설명
PortfolioEnv에서 사용할 보상함수 3종을 구현한다.

## 필요한 이유
프로젝트 요구사항상 단순 수익률, 샤프 기반, MDD 페널티 보상함수를 비교해야 하며, 이후 실험 리포트의 보상함수 실험 근거가 된다.

## 구현 방향 (선택)
- 단순 수익률: reward = r_t
- 샤프 기반: reward = r_t / sigma_t
- MDD 페널티: reward = r_t - lambda * MDD_t
- lambda 기본값은 1.0, 추후 0.5~5.0 범위 실험 가능하게 설정값 분리

## 관련 파일/모듈 (선택)
- `src/rl/env.py`
- `config.py`

## 담당자
- [ ] 강유영 (RAG/프론트)
- [ ] 박지민 (데이터/백엔드)
- [x] 이문정 (RL/인프라)

### LINKED ISSUE #14: [FEAT] PPO 50k 테스트 학습 파이프라인 구현
## 관련 스프린트
- [ ] Sprint 1 (3/31~4/13)
- [x] Sprint 2 (4/14~4/27)
- [ ] Sprint 3 (4/28~5/11)
- [ ] Sprint 4 (5/12~5/25)
- [ ] Sprint 5 (5/26~6/15)

## 기능 설명
Stable-Baselines3 PPO를 사용해 PortfolioEnv에서 50k 스텝 테스트 학습을 수행한다.

## 필요한 이유
Sprint 3의 100k 풀 학습 전에 환경이 정상적으로 학습 가능한지, reward가 NaN 없이 기록되는지, 학습 곡선이 생성되는지 검증해야 한다.

## 구현 방향 (선택)
- `PPO("MlpPolicy", env, tensorboard_log=...)` 형태로 학습
- 50,000 timesteps 우선 실행
- 학습 모델은 `models/ppo_test.zip` 형태로 저장
- TensorBoard 로그 저장 경로 지정
- 학습 실패 시 원인 기록: 관측 shape, reward NaN, done 처리, 데이터 길이 부족 등

## 관련 파일/모듈 (선택)
- `src/rl/train.py`
- `models/`
- `runs/` 또는 `logs/tensorboard/`

## 담당자
- [ ] 강유영 (RAG/프론트)
- [ ] 박지민 (데이터/백엔드)
- [x] 이문정 (RL/인프라)

### PR DESCRIPTION
**Pull request by Dlans00**
## 관련 이슈
closes #12, closes #13, closes #14

## 작업 내용
이문정(m) 담당 RL 환경 핵심 구현 및 PPO 50k 테스트 학습 — Sprint 2 범위.

- `src/rl/env.py`: `PortfolioEnv` 거래 환경 완전 구현
- 거래비용 0.015% + 슬리피지 0.05% 반영
- MDD 15% 초과 시 Safe-Guard 조기 종료 로직 구현
- 보상함수 3종 구현
  - 단순 수익률 기반
  - 샤프 기반
  - MDD 페널티 기반
- `tests/test_env.py`: PortfolioEnv 단위 테스트 추가
- `src/rl/train.py`: PPO 50k 스텝 테스트 학습 스크립트 구현
- 실데이터 parquet 기반 PPO 학습 정상 완료 확인

## 변경된 파일
| 파일 | 변경 내용 |
|------|---------|
| `src/rl/env.py` | PortfolioEnv 거래비용·슬리피지·MDD Safe-Guard·보상함수 3종 구현 |
| `tests/test_env.py` | reset/step, action 정규화, 거래비용, Safe-Guard, 보상함수 3종 테스트 추가 |
| `src/rl/train.py` | `returns.parquet`, `features.parquet` 기반 PPO 50k 테스트 학습 스크립트 추가 |

### PortfolioEnv 설계
- **데이터 정렬**: `features_df` 기준 날짜와 `returns_df` 날짜를 `common_index`로 정렬
- **관찰 공간**: `lookback × n_assets` 정규화 수익률 윈도우 + 현재 포트폴리오 비중 + RSI + MACD_signal
- **관찰 차원**: `(lookback + 3) × n_assets`
  - 현재 기준: `lookback=30`, `n_assets=10` → `(330,)`
- **행동 공간**: 자산별 비중 `[0.0, 1.0]` → 내부에서 합계 1이 되도록 정규화
- **수익률 계산**: 실제 포트폴리오 수익률은 raw 로그수익률인 `returns_df` 기준으로 계산
- **관측값 구성**: 학습 입력은 정규화된 `features_df`의 `{ticker}_return`, `{ticker}_RSI`, `{ticker}_MACD_signal` 사용
- **MACD 처리**: updated SPEC 기준에 맞춰 MACD 원값은 observation에서 제외하고 MACD_signal만 사용

### 거래비용 및 슬리피지
- 거래비용: `0.00015` (0.015%)
- 슬리피지: `0.0005` (0.05%)
- 리밸런싱 시 이전 비중과 새 비중의 차이(`turnover`)에 비례하여 차감

### Safe-Guard
- 포트폴리오 가치 기준 peak 대비 현재 낙폭(MDD)을 계산
- `MDD > 0.15`이면 `terminated=True`
- `info["safe_guard_triggered"]`로 발동 여부 반환

### 보상함수 3종
- `reward_type="return"`
  - `reward = net_return`
- `reward_type="sharpe"`
  - `reward = net_return / volatility`
  - 최근 `volatility_window` 구간의 포트폴리오 수익률 표준편차 사용
- `reward_type="mdd"`
  - `reward = net_return - lambda_mdd * current_mdd`
  - `lambda_mdd` 기본값은 `1.0`

### PPO 50k 테스트 학습
- `src/rl/train.py`에서 실데이터 parquet 로드
  - `data/processed/returns.parquet`
  - `data/processed/features.parquet`
- `PortfolioEnv` 생성 설정
  - `lookback=30`
  - `reward_type="return"`
- PPO 설정
  - `policy="MlpPolicy"`
  - `learning_rate=3e-4`
  - `n_steps=2048`
  - `batch_size=64`
  - `n_epochs=10`
  - `total_timesteps=50_000`
- 학습 완료 후 모델 저장
  - `models/ppo_return_50k.zip`

## 테스트 결과
PPO 테스트 학습 실행 명령어 "python -m src.rl.train" 결과
PPO50k테스트1
image

## 체크리스트
- [x] 담당 파일 외 다른 팀원 파일 무단 수정 없음
- [x] `.env` 파일 커밋 안 함
- [x] 새 함수/클래스에 docstring 작성
- [x] 테스트 통과 확인 (`pytest`)
- [x] PR 대상 브랜치가 `dev`인지 확인 (셀프 merge 금지)

## 기타 참고 사항
- PPO 입력 observation shape은 현재 `lookback=30`, `n_assets=10` 기준 `(330,)`입니다.
- observation 순서는 `[30일 return window → current weights → RSI → MACD_signal]`입니다.
- MACD 원값은 updated SPEC에 맞춰 observation에서 제외하고 `MACD_signal`만 사용했습니다.
- SHAP `/explain` 연동은 아직 구현 전이며 Sprint 3에서 별도 작업 예정입니다.
- Docker Compose 기반 API 컨테이너 구동 확인은 아직 진행 전입니다.

### CODE REVIEW

**Review by uuyeong — COMMENTED** [review]
## 코드 리뷰 — PR #17 PortfolioEnv 구현

---

### 🔴 [Critical] `src/rl/env.py` — 로그수익률 복리 계산 오류

**위치**: `step()` 내 `portfolio_value` 업데이트

```python
# 현재 코드 (잘못됨)
self.portfolio_value *= 1.0 + net_return

# 수정 필요
self.portfolio_value *= np.exp(net_return)
```

`returns_df`는 로그수익률임. 로그수익률은 `exp(r)`로 복리 계산해야 함. `1 + r` 근사는 수익률이 작을 때는 허용 범위지만, 시뮬레이션이 누적될수록 오차가 커지고 **MDD 계산과 Safe-Guard 발동 기준이 틀어짐**. 이 버그가 수정되지 않으면 reward 신호가 실제 수익률을 제대로 반영하지 못함.

`returns_df`가 로그수익률인지 단순수익률인지 `collector.py`랑 크로스체크해서 맞는 방식으로 수정 필요함.

---

### 🟡 [Major] `requirements.txt` — `tensorboard` 버전 미고정

```diff
- tensorboard
+ tensorboard==2.19.0   # 현재 설치 버전으로 고정
```

다른 의존성은 전부 버전 고정돼 있는데 `tensorboard`만 빠져 있음. CI/다른 팀원 환경에서 버전이 달라질 수 있음.

---

### 🟡 [Major] `.github/PULL_REQUEST_m_rl-env_sprint1.md` — 의도치 않은 파일 커밋

PR 템플릿 파일이 실수로 커밋된 것으로 보임. 리포에 있으면 안 되는 파일이므로 삭제 커밋 추가 필요함.

---

### 🟡 [Major] `src/rl/env.py` — 핵심 메서드 docstring 없음

`CLAUDE.md` 컨벤션에 Google 스타일 docstring이 명시돼 있음. `PortfolioEnv.__init__`, `step()`, `reset()`, `_calculate_reward()`에 docstring이 없음. 추가 필요함.

```python
def step(self, action):
    """한 스텝 실행.

    Args:
        action: shape (n_assets,) 비중 벡터. 내부에서 정규화됨.

    Returns:
        obs, reward, terminated, truncated, info 튜플.
    """
```

---

### 🔵 [Minor] `tests/test_env.py` — `test_transaction_cost_is_applied` reward_type 미명시

```python
# 현재
env = PortfolioEnv(returns_df, features_df, lookback=30)

# 권장
env = PortfolioEnv(returns_df, features_df, lookback=30, reward_type="return")
```

`reward == net_return` 단언이 `reward_type="return"`일 때만 성립함. 기본값에 의존하고 있어 기본값 변경 시 테스트가 의도와 다르게 실패할 수 있음. reward_type 명시 권장함.

---

### 🔵 [Minor] `src/rl/env.py` — Sharpe 보상 비연율화 주석 부재

```python
# step-level Sharpe — 연율화하지 않음 (RL 보상 신호용)
return net_return / volatility
```

RL 학습 보상으로 step-level Sharpe를 쓰는 건 맞음. 다만 나중에 `metrics.py`의 연율화된 Sharpe와 해석이 혼동될 수 있으므로 주석 한 줄 추가 권장함.

---

### ✅ 잘된 점

- `common_index`로 날짜 정렬 — features/returns 불일치 방어 로직 탄탄함
- `_normalize_action()` 분리, zero-sum 엣지케이스 처리 깔끔함
- `info` dict에 `safe_guard_triggered`, `gross_return`, `transaction_cost` 등 디버그 정보 풍부함
- 테스트 11개, 보상함수 3종 각각 검증 + 엣지케이스(zero action, invalid type) 커버함
- `model.save("models/ppo_return_50k")` 경로가 스펙과 일치함

File: src/rl/env.py [hunk 1]
```diff
@@ -34,56 +66,117 @@ def __init__(self, returns_df, features_df, lookback=30):
...
+        gross_return = float(np.dot(new_weights, asset_returns))
 
-        asset_returns = self.returns_df.iloc[self.current_step].values.astype(np.float32)
-        reward = float(np.dot(self.weights, asset_returns))
+        turnover = float(np.sum(np.abs(new_weights - self.weights)))
+        transaction_cost = turnover * self.total_cost_rate
 
+        net_return = gross_return - transaction_cost
+
+        self.portfolio_value *= 1.0 + net_return
```
Comment (uuyeong) [inline src/rl/env.py]: `returns_df`는 로그수익률이므로 `1 + r` 근사가 아닌 `exp(r)`로 복리 계산해야 함. 시뮬레이션이 누적될수록 오차가 커지고 MDD 계산과 Safe-Guard 발동 기준이 틀어짐. 이 버그가 수정되지 않으면 reward 신호가 실제 수익률을 제대로 반영하지 못함.

```python
# 현재 (잘못됨)
self.portfolio_value *= 1.0 + net_return

# 수정 필요
self.portfolio_value *= np.exp(net_return)
```

`returns_df`가 로그수익률인지 단순수익률인지 `collector.py`랑 크로스체크해서 맞는 방식으로 수정 필요함.

File: src/rl/env.py [hunk 2]
```diff
@@ -4,20 +4,52 @@
 
 
 class PortfolioEnv(gym.Env):
-    def __init__(self, returns_df, features_df, lookback=30):
+    def __init__(
```
Comment (uuyeong) [inline src/rl/env.py:12]: 핵심 메서드 docstring 없음
`CLAUDE.md` 컨벤션에 Google 스타일 docstring이 명시돼 있음. `__init__`, `step()`, `reset()`, `_calculate_reward()` 모두 docstring 추가 필요함.

File: src/rl/env.py [hunk 3]
```diff
@@ -34,56 +66,117 @@ def __init__(self, returns_df, features_df, lookback=30):
...
+            start = max(0, self.current_step - self.volatility_window)
+            recent_returns = self.returns_df.iloc[start:self.current_step].values
+
+            portfolio_returns = recent_returns @ weights
+            volatility = float(np.std(portfolio_returns))
+
+            if volatility < 1e-8:
+                return 0.0
+
+            return net_return / volatility
```
Comment (uuyeong) [inline src/rl/env.py:164]: RL 보상 신호로 step-level Sharpe를 사용하는 건 맞음. 다만 나중에 `metrics.py`의 연율화된 Sharpe와 해석이 혼동될 수 있으므로 주석 한 줄 추가 권장함.

File: requirements.txt [hunk 4]
```diff
@@ -34,3 +34,4 @@ curl_cffi==0.11.1
 # RL
 gymnasium[classic-control]==1.2.3
 stable-baselines3==2.8.0
+tensorboard
```
Comment (uuyeong) [inline requirements.txt]: 다른 의존성은 전부 버전 고정돼 있는데 `tensorboard`만 빠져 있음. CI/다른 팀원 환경에서 버전이 달라질 수 있음.

File: .github/PULL_REQUEST_m_rl-env_sprint1.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
+## 관련 이슈
```
Comment (uuyeong) [inline .github/PULL_REQUEST_m_rl-env_sprint1.md]: Sprint 1 PR 작성용 템플릿 파일이 실수로 리포에 커밋된 것으로 보임. `.github/` 디렉터리에 있으면 안 되는 파일이므로 삭제 커밋 추가 필요함.

File: tests/test_env.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+
+    expected_weights = np.ones(len(tickers), dtype=np.float32) / len(tickers)
+
+    assert np.allclose(info["weights"], expected_weights)
+
+
+def test_transaction_cost_is_applied(sample_data):
+    returns_df, features_df, _ = sample_data
+
+    env = PortfolioEnv(returns_df, features_df, lookback=30)
```
Comment (uuyeong) [inline tests/test_env.py]: `reward == net_return` 단언이 `reward_type="return"`일 때만 성립함. 현재 기본값에 의존하고 있어 기본값 변경 시 테스트가 의도와 다르게 실패할 수 있음.

File: src/rl/env.py (hunk 1 above)
Comment (Dlans00) [reply src/rl/env.py]: 확인했습니다. 데이터 파이프라인에서 returns_df가 로그수익률로 생성되는 구조라서, portfolio_value 갱신을 1 + r 근사가 아니라 np.exp(r) 기반 복리 계산으로 수정하겠습니다. MDD 및 Safe-Guard 기준도 해당 portfolio_value 기준으로 일관되게 계산되도록 반영하겠습니다.

File: src/rl/env.py (hunk 2 above)
Comment (Dlans00) [reply src/rl/env.py:12]: 반영하겠습니다. CLAUDE.md 컨벤션에 맞춰 __init__, reset, step, _calculate_reward에 Google 스타일 docstring을 추가하겠습니다.

File: requirements.txt (hunk 4 above)
Comment (Dlans00) [reply requirements.txt]: Docker/CI 환경 재현성을 위해 tensorboard도 명시적으로 버전 고정하겠습니다.

File: .github/PULL_REQUEST_m_rl-env_sprint1.md (hunk 5 above)
Comment (Dlans00) [reply .github/PULL_REQUEST_m_rl-env_sprint1.md]: 확인했습니다. Sprint 1 PR 작성용 임시 템플릿이 .github/에 포함된 것은 의도한 변경이 아니므로 삭제하겠습니다.

File: src/rl/env.py (hunk 3 above)
Comment (Dlans00) [reply src/rl/env.py:164]: Sharpe 보상은 학습용 step-level reward이고, metrics.py의 연율화 Sharpe와 목적이 다르다는 주석을 추가하겠습니다.

File: tests/test_env.py (hunk 6 above)
Comment (Dlans00) [reply tests/test_env.py]: 맞습니다. 해당 단언은 reward_type="return"에서만 성립하므로 테스트 fixture 또는 env 생성부에 reward_type을 명시하겠습니다.
